### ISSUE TITLE
Using Scroll Reveal in a React Component

### ISSUE THREAD

**Issue by joshleong** [jlmakes/scrollreveal.js#218]
Been attempting to use ScrollReveal in a react component, everything seems to be working except ScrollReveal isn't being triggered on scrolling, but re-sizing the window in order to hide the elements.

**Comment by jlmakes** [comment]
Hey @joshleong 

I’ve spent a little time with React, but not enough to offer a solution off the top. You’re saying scroll events do nothing, but resize events work as expected?

**Comment by joshleong** [comment]
Ah, cool let me fiddle with that for a bit, and I'll put together a screencast with more details together.

**Comment by joshleong** [comment]
Hey Julian, I put together a screencast with a little more details https://youtu.be/LpoesOwcTw8. 

Journal Component:
http://jsbin.com/jisazihabi/edit?html,output

SiteLayout Component:
http://jsbin.com/xucohabixi/edit?html,output

**Comment by jlmakes** [comment]
Thanks for the video @joshleong :bow:

If resize works correctly—that means the visibility logic, config and style factories are working...

I’m taking a shot in the dark as I don’t know your markup, but it looks like you were attempting to use a custom container—but never actually used it in your ScrollReveal configuration.

```diff
afterMount = () => {

- const sr = ScrollReveal();
  const container = document.querySelector('.grid-journal');
+ const sr = ScrollReveal({ container: container });

  console.log('AfterMount');

  const logger = () => {
    logviewport();
  };

  //setInterval(logger, this.interval);

  toggleVideo('video-stacking');
  sr.reveal('.header-video', {

              duration: 500,
              distance: '0px',
              easing: 'ease-out',
              scale: 1.1,
              reset: true,
              afterReveal: function() {toggleVideo('video-stacking'); console.log('AfterRevealFired');},
              afterReset: function() {toggleVideo('video-stacking'); console.log('AfterResetFired');}

          });

  sr.reveal('.body-information', {

              duration: 500,
              distance: '0px',
              easing: 'ease-out',
              scale: 1.1,
              reset: true,
              afterReveal: function() {console.log('Body-AfterRevealFired');},
              afterReset: function() {console.log('Body-AfterResetFired');}

          });

}
```

Let me know if I’m on to something.

Also, I couldn’t help but read a little of what was on your screen—totally stole a sneak peak at the product you’re working on. I would love to hear a little more about it if you’re willing.

**Comment by joshleong** [comment]
Yes that worked!

**Comment by joshleong** [comment]
Hey Thanks! That worked!

Oh Grid, this old thing, be happy to chat about it. I'm free this week, I'm
in San Francisco, where are you located?

Best,
Josh

**Comment by oyeanuj** [comment]
@joshleong What was your final solution of integrating ScrollReveal in React? Possible to provide a gist/example?

**Comment by joshleong** [comment]
Actually Julian's answer made my code work, it's up there in the conversation!

**Comment by 503945930** [comment]
i want to read success code. thanks

**Comment by ArnaudValensi** [comment]
Here is a full example:

`npm install --save scrollreveal` or install like you're used to doing it.

It doesn't work well if there are multiple instances of ScrollReveal, so we have to create a module returning an instance:

`file scrollReveal.js`:
```
import ScrollReveal from 'scrollreveal'

export default ScrollReveal()
```

Then in a component:
```
import React from 'react'
import sr from './scrollReveal.js'

export class Testimonial extends React.Component {
  props: Props;

  componentDidMount = () => {
    const config = {
      origin: 'right',
      duration: 1000,
      delay: 150,
      distance: '500px',
      scale: 1,
      easing: 'ease',
    }

    sr.reveal(this.refs.box1, config)
  }

  render () {
    return (
      <section className='testimonial' id='testimonials'>
        <div className='row' ref='box1'>
          ...
        </div>
      </section>
    )
  }
}

export default Testimonial
```

We configure scrollreveal in componentDidMount. We access the DOM element, setting a `ref='name'` and accessing using `this.refs.name`.

Last thing, the entire page must scroll, il you have a container element which does an `overflow: hidden` for example, the animation may not trigger. I think you can check it doing this:
`document.onscroll = function() { console.log('Scroll ok); };`
Maybe in this case we can set the scrolling container, I don't know.

That's it, it works very well this way.

**Comment by renatobenks** [comment]
Hey guys and @jlmakes, someone have a idea in how cover with unit tests the scrollreveal uses in component react with [Jest](https://facebook.github.io/jest/) with environment Node, and `react-test-renderer` lib to shoot snapshots? (Not using enzyme or other lib within).

I'm using document, window, etc, mocked to test scrollreveal behaviour, what already cost me too much hard works and time. So, my issue it's only make the scrollreveal works fine.

Here is on my follow component to test
```javascript
class ScrollAnimate extends Component {
    static getOffset ({top, left, right, bottom}) {
        return {
            top: top || 0,
            left: left || 0,
            right: right || 0,
            bottom: bottom || 0
        }
    }

    constructor (props) {
        super(props);
        this.state = {
            scroll: new ScrollReveal({
                scale: !!props.scale ? props.scale === true ? 0.1 : props.scale : null,
                duration: props.duration || 1500,
                reset: !!props.reset,
                viewFactor: props.view || 0.2,
                viewOffset: props.offset
                    ? ScrollAnimate.getOffset(props.offset)
                    : ScrollAnimate.getOffset({})
            }, 50)
        }
    }

    componentDidMount () {
        const { scroll } = this.state;
        const { loadAfter, children } = this.props;
        scroll.reveal(String.prototype.concat('#', children.props.id), {
            afterReveal () {
                return loadAfter();
            }
        });
    }

    render () {
        return this.props.children;
    }
}
```

**Comment by jlmakes** [comment]
### ScrollReveal 4
As of version `4.0.0-beta.10`, the ScrollReveal constructor returns a singleton—which means the extra file used in the example above is no longer necessary:

~_Create new file: `scrollReveal.js`:_~

```diff
- import ScrollReveal from 'scrollreveal'

- export default ScrollReveal()
```

Then in a component:

``` diff
  import React from 'react'
- import sr from './scrollReveal.js'
+ import ScrollReveal from 'scrollreveal'

  export class Testimonial extends React.Component {
    props: Props;

    componentDidMount = () => {
      const config = {
        origin: 'right',
        duration: 1000,
        delay: 150,
        distance: '500px',
        scale: 1,
        easing: 'ease',
      }

-     sr.reveal(this.refs.box1, config)
+     ScrollReveal().reveal(this.refs.box1, config)
    }

    render () {
      return (
        <section className='testimonial' id='testimonials'>
          <div className='row' ref='box1'>
            ...
          </div>
        </section>
      )
    }
  }

  export default Testimonial
```

Some official documentation on ScrollReveal + React integration is coming soon!

**Comment by sniebauer** [comment]
Any news on Scroll Reveal React? Stoked to try it out!

**Comment by jlmakes** [comment]
@sniebauer I appreciate the sentiment!

The latest `v4.0.0-beta.14` has been a stable release candidate for a couple weeks now, so I will launch as soon as I finish the new website/docs.

This late in the game, I don’t plan on holding up launch for React specific documentation, but it is a priority for me: I want to serve more intermediate/advanced developers, and (to be blunt), piggy back off the React hype train.

The idea of a more complicated integration—such as a separate package that provides a higher-order component—_is_ interesting to me, but to be honest, I’m not yet sure what constraints I’ll run into trying to express the ScrollReveal 4 API while Thinking in React™ (or if a separate package/higher-order component is even necessary/valuable, and maintainable).

Realistically, I think simple live integration examples (similar to what’s [shown](https://github.com/jlmakes/scrollreveal/issues/218#issuecomment-229907778) [above](https://github.com/jlmakes/scrollreveal/issues/218#issuecomment-321856891)) will be included at (or shortly after) launch. The rest, depends on what I uncover.

**Comment by jelenajjo** [comment]
@jlmakes  Any updates? :) It is confusing and hard to use this library in React apps at this point, documentation and a bit better support would really mean a lot. 

Especially with **Sequenced Animations**, because in React we have just one component that we pass multiple times, so we have same `ref` on all those elements. And then only the last elements gets to be animated.

I love this library, I use it for all my projects, like:

- https://vanila.io
- https://automatio.co

I saw few other similar libraries/packages made for React but I would really like to stick to this one since I use it for so long and it provides the best experience. I get positive feedback on daily basis cause of those animations. 

If you can make this in top few priority things on list that would be awesome. :+1:

**Comment by jlmakes** [comment]
I appreciate the support @jelenajjo 

Nothing new in regards to React to report yet, still finishing up the new docs.

The official React docs detail [Integrating with DOM Manipulation Plugins](https://reactjs.org/docs/integrating-with-other-libraries.html#integrating-with-dom-manipulation-plugins), which will likely be my guide when I have the time; have you gone through that yet?

**Comment by jelenajjo** [comment]
I will take a look again! :) Okay, looking forward to finished docs. :-1:

**Comment by jlmakes** [comment]
I'm experimenting unofficially right now @jelenajjo, if I find a good solution I’ll send it your way.

**Comment by jlmakes** [comment]
### ⚠️ Disclaimer: Work In Progress!
Since you brought this up, I thought I'd at least take a stab at a React ScrollReveal integration proof of concept; this is what I’ve come up with so far:

```jsx
// WithReveal.js

import React, { Component } from 'react'
import { findDOMNode } from 'react-dom'
import ScrollReveal from 'scrollreveal'

export const WithReveal = Original =>
  class extends Component {
    constructor(props) {
      super(props)
      this.target = []
    }
    componentDidMount() {
      ScrollReveal().reveal(
        this.target,
        this.props.options,
        this.props.interval
      )
    }
    componentWillUnmount() {
      ScrollReveal().clean(this.target)
    }
    render() {
      const children = React.Children.map(this.props.children, child =>
        React.cloneElement(child, {
          ref: c => this.target.push(findDOMNode(c))
        })
      )
      return <Original {...this.props}>{children}</Original>
    }
  }

export default WithReveal
```

It’s a pretty small higher order component, that simply captures the DOM nodes of the `Original` component’s children—and then passes them to ScrollReveal via lifecycle methods.

Imagine we have a simple component like this:

```jsx
const List = ({ children }) => (
  <div>
    <h1>List:</h1>
    {children}
  </div>
)
```

All we'd have to do is import the ScrollReveal HOC and wrap it:

```jsx
// List.js

import React from 'react'
import WithReveal from './WithReveal.js'

const List = ({ children }) => (
  <div>
    <h1>List:</h1>
    {children}
  </div>
)

export default WithReveal(List)
```

Then in our app, we might use it like this:

```jsx
import React from 'react'
import { render } from 'react-dom'

import List from './List'

const App = () => (
  <List options={{ duration: 2000 }} interval={200}>
    <div>foo</div>
    <div>bar</div>
    <div>jam</div>
  </List>
)

render(<App />, document.getElementById('root'))
```

I’m sure there's room for improvement (or even bugs), you may want to organize your files differently, and the `options` and `interval` props might be too generic and cause naming collisions.

Hopefully though, this at least helps people trying to work with React and ScrollReveal for the time being. Check out the [Live Example on CodeSandbox](https://codesandbox.io/s/z2wqk3vm1l)

**Comment by jelenajjo** [comment]
@jlmakes  Thanks! I will test this out now, and if I come up with improvements I will share def.

**Comment by jlmakes** [comment]
Also, I forgot to mention—since we're relying on references, stateless functional components do not work as reveal targets. Take this example:

```js
// Item.js

import React, { Component } from 'react'

const Item = ({ content }) => <li>{content}</li>

export default Item
```

That means if we tried to use them like this:

```js
const App = () => (
  <div>
    <RevealedList options={{ distance: '50px', origin: 'right' }} interval={500}>
      <Item content="foo" />
      <Item content="bar" />
      <Item content="jam" />
    </RevealedList>
  </div>
)
```
We won't get any action cause React just doesn't expose references for stateless functional components. Even though it's a crazy simple, we still would need to use a the component class:

```js
import React, { Component } from 'react'

class Item extends Component {
  render() {
    return <li>{this.props.content}</li>
  }
}

export default Item

```

I’ve updated the [CodeSandbox](https://codesandbox.io/s/z2wqk3vm1l) to reflect this change.

**Comment by ryanturnbull** [comment]
any update on the react version @jlmakes ?

**Comment by jlmakes** [comment]
As it stands @ryanturnbull, the higher-order component I outlined above works well—I'm mainly just looking for more feedback on the implementation. It's only about 30 lines of code, so I'm not sure it warrants a separate package for now. (Plus this way, you can manage prop naming conflicts.)

Another developer has created `react-scrollreveal` but @RusinovAnton has a different opinion about what the integration API should look like. It's also worth mentioning that his package is only available under the GPLv3 (as stipulated by me), so commercial use it out of the question.

It's possible he and I will work together to make that an official package (Hi Anton!), but for now I recommend using the HOC detailed [above](https://github.com/jlmakes/scrollreveal/issues/218#issuecomment-341253413) and sharing your feedback.

**Comment by RusinovAnton** [comment]
@jlmakes Hey, Julian! I would love to help with official react-scrollreveal component. Currently I don't have much time for that though, but I believe I may get to it in two weeks or so.

I want to share some of my findings while developing react-scrollreveal. That component initially was made for certain project where I had to use ScrollReveal with react. At first it was separate components for each animated element but I've faced performance issues so after that I made HOC that intended to wrap whole page on the highest level in components tree and then I declared all animated elements on that page by passing needed selectors, then on component's mount single ScrollReveal instance was created and managed needed animations. 

When I'll have time I would look closely at HOC that you've suggested here and open issue with implementation draft in react-scrollreveal repo. Feel free to open it yourself if you want to, I think that would be a good start of work and we could move discussion there.
